hierarchicalforecast/core.py

```diff
@@ -97,99 +103,142 @@ def __init__(self,
 
         # Declare output names
-        drop_cols = ['ds', 'y'] if 'y' in Y_hat_df.columns else ['ds']
-        model_names = Y_hat_df.drop(columns=drop_cols, axis=1).columns.to_list()
+        model_names = list(set(Y_hat_df_nw.columns) - set([id_col, time_col, target_col]))
 
         # Ensure numeric columns
-        if not len(Y_hat_df[model_names].select_dtypes(include='number').columns) == len(Y_hat_df[model_names].columns):
-            raise Exception('`Y_hat_df`s columns contain non numeric types')
-            
-        #Ensure no null values
-        if Y_hat_df[model_names].isnull().values.any():
-            raise Exception('`Y_hat_df` contains null values')
-        
-        pi_model_names = [name for name in model_names if ('-lo' in name or '-hi' in name or '-median' in name)]
-        model_names = [name for name in model_names if name not in pi_model_names]
+        assert Y_hat_df_nw[model_names].columns == Y_hat_df_nw[model_names].select(nw.selectors.numeric()).columns, "All models in `Y_hat_df` must contain numeric values."
+
+        # Ensure no null values
+        assert not Y_hat_df_nw[model_names].select(nw.all().is_null().any()).to_numpy().any(), "`Y_hat_df` contains null values. Make sure no column in `Y_hat_df` contains null values."
         
         # TODO: Complete y_hat_insample protection
-        if intervals_method in ['bootstrap', 'permbu'] and Y_df is not None:
-            if not (set(model_names) <= set(Y_df.columns)):
-                raise Exception('Check `Y_hat_df`s models are included in `Y_df` columns')
+        model_names = [name for name in model_names if not ('-lo' in name or '-hi' in name or '-median' in name)]        
+        if intervals_method in ['bootstrap', 'permbu'] and Y_df_nw is not None:
+            if not (set(model_names) <= set(Y_df_nw.columns)):
+                raise Exception(f"Check `Y_df` columns, {model_names} must be in `Y_df` columns.")
 
-        uids = Y_hat_df.index.unique()
+        # Assert S is an identity matrix at the bottom
+        S_np = S_df_nw.drop(id_col).to_numpy()
+        if not np.allclose(S_np[-S_np.shape[1]:], np.eye(S_np.shape[1])):
+            raise ValueError(f"The bottom {S_np.shape[1]}x{S_np.shape[1]} part of S must be an identity matrix.")
 
         # Check Y_hat_df\S_df series difference
-        S_diff  = len(S_df.index.difference(uids))
-        Y_hat_diff = len(Y_hat_df.index.difference(S_df.index.unique()))
-        if S_diff > 0 or Y_hat_diff > 0:
-            raise Exception(f'Check `S_df`, `Y_hat_df` series difference, S\Y_hat={S_diff}, Y_hat\S={Y_hat_diff}')
+        # TODO: this logic should be method specific
+        S_diff = set(S_df_nw[id_col]) - set(Y_hat_df_nw[id_col])
+        Y_hat_diff = set(Y_hat_df_nw[id_col]) - set(S_df_nw[id_col])
+        if S_diff:
+            raise Exception(f'There are unique_ids in S_df that are not in Y_hat_df: {S_diff}')
```

Comment:
These should be ValueErrors

Reply:
I love these detailed comments, thanks 👍